openml/runs/trace.py

```diff
@@ -44,35 +46,193 @@ def get_selected_iteration(self, fold, repeat):
+            tit_list = []
+            for attr, attr_type in trace_attributes:
+                value = getattr(trace_iteration, attr)
+                if attr == 'selected':
+                    if value:
+                        tit_list.append('true')
+                    else:
+                        tit_list.append('false')
+                else:
+                    tit_list.append(value)
+            data.append(tit_list)
+
+        arff_dict['attributes'] = trace_attributes
+        arff_dict['data'] = data
+        arff_dict['relation'] = "Trace"
+
+        return arff_dict
+
+    @classmethod
+    def trace_from_arff(cls, arff_obj):
+        """Generate trace from arff trace.
+
+        Creates a trace file from arff object (for example, generated by a local run).
+
+        Parameters
+        ----------
+        arff_obj : dict
+            LIAC arff obj, dict containing attributes, relation, data.
+
+        Returns
+        -------
+        OpenMLRunTrace
+        """
+        trace = OrderedDict()
+        # flag if setup string is in attributes
+        flag_ss = False
+        attribute_idx = {att[0]: idx for idx, att in enumerate(arff_obj['attributes'])}
+        for required_attribute in ['repeat', 'fold', 'iteration', 'evaluation', 'selected']:
+            if required_attribute not in attribute_idx:
+                raise ValueError('arff misses required attribute: %s' % required_attribute)
+        if 'setup_string' in attribute_idx:
+            flag_ss = True
+
+        for itt in arff_obj['data']:
+            repeat = int(itt[attribute_idx['repeat']])
+            fold = int(itt[attribute_idx['fold']])
+            iteration = int(itt[attribute_idx['iteration']])
+            evaluation = float(itt[attribute_idx['evaluation']])
+            setup_string = None
+            if flag_ss:
+                if itt[attribute_idx['setup_string']] != None:
+                    setup_string = str(itt[attribute_idx['setup_string']])
+            selected_value = itt[attribute_idx['selected']]
+            if selected_value == 'true':
+                selected = True
+            elif selected_value == 'false':
+                selected = False
+            else:
+                raise ValueError('expected {"true", "false"} value for selected field, received: %s' % selected_value)
+
+            # if someone needs it, he can use the parameter
+            # fields to revive the setup_string as well
+            # However, this is usually done by the OpenML server
+            # and if we are going to duplicate this functionality
+            # it needs proper testing
+
+            current = OpenMLTraceIteration(repeat, fold, iteration, setup_string,
+                                           evaluation, selected)
+            trace[(repeat, fold, iteration)] = current
+
+        return cls(None, trace)
+
+    @classmethod
+    def _trace_from_xml(cls, xml):
+        """Generate trace from xml.
```

Comment:
Generalizing is not easy, since they use different attribute names and also one is a list while the other one is a dictionary. Skipping it for the moment.